### DISCUSSION TITLE
Should C# support mixing deconstructors and target-typed switch/option expressions?

### DISCUSSION THREAD

**Discussion by DavidArno** [dotnet/csharplang#6953] [Language Ideas]
I can use a deconstruct with a switch statement if the compiler can infer the expression type from the expression arms:
```cs
// compiles
(int x, int y) = true switch {
    true => (1, 1),
    false => (2, 2)
};
```
I use a tuple as the target-type if the compiler can infer the expression type from that target type:
```cs
// compiles
(Func<int>, int) t = true switch {
        true => (() => 1, 1),
        false => (() => 2, 2)
};
```
I can use a deconstruct with a switch statement, as long as I also supply a variable, so that the compiler can use the variable's target type to infer the two of the expression and thus decompose that variable:
```cs
// compiles
(Func<int> v, int w) z = true switch {
        true => (() => 1, 1),
        false => (() => 2, 2)
};
```
But if I skip the variable, the compiler cannot infer the target type from the deconstruct:
```cs
// Doesn't compile: CS8506 and CS8131 errors
(Func<int> v, int w) = true switch {
        true => (() => 1, 1),
        false => (() => 2, 2)
};
```

It would be good therefore to adjust the compiler's behaviour here and have it lower the above to add a "unspeakable" variable and then apply the target-typing rules to that:
```cs
(Func<int> v, int w) = true switch {
        true => (() => 1, 1),
        false => (() => 2, 2)
};
// gets lowered to:
(Func<int> v, int w) <>_tmp_1234 = true switch {
        true => (() => 1, 1),
        false => (() => 2, 2)
};
// and so compiles
```
The same situation (and the same solution) could be applied to optional expressions too:
```cs
(int x, int y) = true ? (1, 1) : (2, 2); // compiles
(Func<int>, int) t = true ? (() => 1, 1) : (() => 2, 2); // compiles
(Func<int> v, int w) z = true ? (() => 1, 1) : (() => 2, 2); // compiles
(Func<int> v, int w) = true ? (() => 1, 1) : (() => 2, 2); // doesn't compile
// so lower it to:
(Func<int> v, int w) <>_tmp_1234 = true ? (() => 1, 1) : (() => 2, 2);
// and it can then compile
```

**Comment by RikkiGibson** [comment]
I find it slightly surprising that a deconstruction assignment `(A a, B b) = c` doesn't provide a similar target type for `c` as `(A a, B b) t = c`. Perhaps it would be worth reviewing the relevant proposals, LDM notes and test plans to confirm the intent there.

**Reply by CyrusNajmabadi** [reply]
That is also surprising for me.  It doesn't feel intentional IMO.  Perhaps just an unfortunate fallout on how we phrased things.  @jcouv do you want to bring a clarification question to LDM here?

**Reply by DavidArno** [reply]
@CyrusNajmabadi, did this get considered for discussion at an LDM?

**Reply by CyrusNajmabadi** [reply]
No.  Up to @jcouv if he'd like to drive this. :)

**Reply by DavidArno** [reply]
OK, thanks.

**Reply by RikkiGibson** [reply]
BTW, anybody is also welcome to contribute to answering this question by searching the "relevant proposals, LDM notes and test plans" :)

If there's evidence of what our intent was one way or the other, that would of course tell us if action is needed here.

**Reply by TheUnlocked** [reply]
I've been searching through every reference to tuples and deconstruction in the 2015-2016 meeting notes and there doesn't seem to be anything which considers this. The best hint I could find is at https://github.com/dotnet/csharplang/blob/main/meetings/2016/LDM-2016-07-13.md#deconstruction-of-tuple-literals, which discusses direct tuple deconstruction.

> It should all work. Even though there never observably is a physical tuple in existence (it can be thought of as a series of point-wise assignments), **semantics should correspond to introducing a fake tuple type, then imposing it on the RHS**. [emphasis added]

Though it's not clear if "imposing it on the RHS" is _just_ referring to implicit conversions or if it was also meant to include target typing. The algorithm described after that paragraph only mentions conversions.

**Reply by jcouv** [reply]
The rules for giving the right-hand-side some type information in deconstructions aren't a simple target-typing.
For example, `(string i, var j) = (null, 42);` works and shows that neither side has a type. I'd have to dig to see how well this was documented, but I think it should be in some LDM notes at least.

Those rules work well with tuples, but don't work for other expressions (like a `switch`). 
I'm not sure how (from a language/spec perspective) to maintain the former while improving the latter.
